larsim/IonizationScintillation/ISCalcNESTLAr.cxx

```diff
@@ -57,53 +57,33 @@ namespace larg4
         }
 
         int pdgcode = edep.PdgCode();
-
-        geo::Length_t startx = edep.StartX();
-        geo::Length_t starty = edep.StartY();
-        geo::Length_t startz = edep.StartZ();
-        geo::Length_t endx   = edep.EndX();
-        geo::Length_t endy   = edep.EndY();
-        geo::Length_t endz   = edep.EndZ();
-//        geo::Point_t  midpos = edep.MidPoint();
-
-//        double delta    = 1*CLHEP::mm;
-//        double R0           = 1.568*CLHEP::um;    //Mozumder 1995
-
+        
         double DokeBirks[3];
-//        double ThomasImel = 0.00;
 
-        double eField = EFieldAtStep(fDetProp->Efield(), edep);
+//        double eField = EFieldAtStep(fDetProp->Efield(), edep);
+        double eField = EFieldAtStep(fDetProp->Efield(), edep) * 1000;  //conver kV to 1000 V
         if(eField)
         {
-//            ThomasImel   = 0.156977 * pow(eField, -0.1);
             DokeBirks[0] = 0.07 * pow((eField/1.0e3), -0.85);
             DokeBirks[2] = 0.00;
         }
         else
         {
-//            ThomasImel   = 0.099;
             DokeBirks[0] = 0.0003;
             DokeBirks[2] = 0.75;
         }
 
-        double Density = fDetProp->Density()/(CLHEP::g/CLHEP::cm3); // argon density at the temperature from Temperature()
+//        double Density = fDetProp->Density()/(CLHEP::g/CLHEP::cm3); // argon density at the temperature from Temperature()
+        double Density = LAr_Density;
```

Comment:
How come you want to drop using the detector properties service here?

Reply:
I have just modified the `DetectorPropertiesStandard` unit test to make it print the LAr density (`lardataalg` pull request lardataalg#14).
`DetectorPropertiesStandardTest standard_g4_icarus.fcl` prints a reasonable 1.39 kg/dm^3^.
So `DetectorProperties` _shouldn't_ be the problem. Passing the service provider through NESTed layers and resetting it on each event might be more of a challenge.

Reply:
This is just more mysterious. I would have assumed the difficult part per @PetrilloAtWork's message would be just _getting_ a service handle. But, once we've gotten the handle, why would it get a nonsense value of the density? Are there instructions for running the unit tests? Perhaps @Healthborn can try running those, and confirm he's seeing different values for the density in the two contexts. It might help track down the issue.

Reply:
The unit test is _art_-independent, and it utilises `DetectorPropertiesStandard` and `LArPropertiesStandard` (and `DetectorClocksStandard` and maybe `Geometry` with standard channel mapping: neither of these two dependencies should matter for the density). If DUNE uses these standard implementations, then the unit test is useful and can be run with `DetectorPropertiesStandardTest standard_g4_dune.fcl` (where the latter is any job configuration loading the relevant `DetectorProperties` configuration). The unit test should be in the `PATH` when `lardataalg` is set up or compiled.
If DUNE is using a custom implementation, the unit test won't exercise it. But I think David Adams also had some way to load services off-_art_, provided that it survived LArSoft `v09`.